### ISSUE TITLE
Plotter crashes when using `show_plot=False, save_plot=True`

### ISSUE THREAD

**Issue by mseitzer** [f-dangel/cockpit#9]
## Description

Plotting to a file crashes (see stacktrace below). Calling plot with `show_plot=True, save_plot=True` works and results in the file being saved correctly. The issue is reproducible over multiple runs and thus probably not caused by weird gradients etc.

## Steps to Reproduce

1. Setup

```
config = [
                quantities.Distance(schedules.linear(interval=1)),
                quantities.GradHist1d(schedules.linear(interval=1)),
                quantities.GradNorm(schedules.linear(interval=1)),
                quantities.InnerTest(schedules.linear(interval=1)),
                quantities.Loss(schedules.linear(interval=1)),
                quantities.NormTest(schedules.linear(interval=1)),
                quantities.OrthoTest(schedules.linear(interval=1)),
                quantities.UpdateSize(schedules.linear(interval=1)),
                quantities.GradHist2d(schedules.linear(interval=1))
            ]
            self._cockpit = Cockpit(model.parameters(), quantities=config)
            self._cockpit_plotter = CockpitPlotter()
```

2. Track a few backward steps

3. Call plot with `show_plot=False` and `save_plot=True`
```
self._cockpit_plotter.plot(self._cockpit,
                                        savedir=cockpit_log_dir,
                                        show_plot=False,
                                        save_plot=True)
```

---

```
Traceback (most recent call last):
  File "code/train.py", line 215, in <module>
    metrics = main(opts)
  File "code/train.py", line 185, in main
    training_info = trainer.train(model, dataset, eval_data, logger)
  File "./code/trainers/mse_trainer.py", line 186, in train
    self.log(logger.root_dir)
  File "./code/trainers/base_trainer.py", line 145, in log
    save_plot=True)
  File "./envs/pytorch/lib/python3.7/site-packages/cockpit/plotter.py", line 142, in plot
    self._plot_gradients(self.grid_spec[0, 1])
  File "./envs/pytorch/lib/python3.7/site-packages/cockpit/plotter.py", line 223, in _plot_gradients
    instruments.gradient_tests_gauge(self, self.fig, self.gs_gradients[1, 1])
  File "./envs/pytorch/lib/python3.7/site-packages/cockpit/instruments/gradient_tests_gauge.py", line 69, in gradient_tests_gauge
    _format(self, ax_all, ax_norm, ax_inner, ax_ortho)
  File "./envs/pytorch/lib/python3.7/site-packages/cockpit/instruments/gradient_tests_gauge.py", line 91, in _format
    ax_norm.set_yscale("log")
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/axes/_base.py", line 3708, in set_yscale
    ax.yaxis._set_scale(value, **kwargs)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/axis.py", line 800, in _set_scale
    self._scale.set_default_locators_and_formatters(self)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/scale.py", line 406, in set_default_locators_and_formatters
    axis.set_major_locator(LogLocator(self.base))
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/axis.py", line 1651, in set_major_locator
    self.stale = True
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/artist.py", line 230, in stale
    self.stale_callback(self, val)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/artist.py", line 51, in _stale_axes_callback
    self.axes.stale = val
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/artist.py", line 230, in stale
    self.stale_callback(self, val)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/figure.py", line 51, in _stale_figure_callback
    self.figure.stale = val
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/artist.py", line 230, in stale
    self.stale_callback(self, val)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/pyplot.py", line 589, in _auto_draw_if_interactive
    fig.canvas.draw_idle()
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/backend_bases.py", line 1907, in draw_idle
    self.draw(*args, **kwargs)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/backends/backend_agg.py", line 388, in draw
    self.figure.draw(self.renderer)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/artist.py", line 38, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/figure.py", line 1709, in draw
    renderer, self, artists, self.suppressComposite)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/image.py", line 135, in _draw_list_compositing_images
    a.draw(renderer)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/artist.py", line 38, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/axes/_base.py", line 2647, in draw
    mimage._draw_list_compositing_images(renderer, self, artists)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/image.py", line 135, in _draw_list_compositing_images
    a.draw(renderer)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/artist.py", line 38, in draw_wrapper
    return draw(artist, renderer, *args, **kwargs)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/axis.py", line 1203, in draw
    ticks_to_draw = self._update_ticks()
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/axis.py", line 1079, in _update_ticks
    major_locs = self.get_majorticklocs()
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/axis.py", line 1324, in get_majorticklocs
    return self.major.locator()
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/ticker.py", line 2271, in __call__
    return self.tick_values(vmin, vmax)
  File "./envs/pytorch/lib/python3.7/site-packages/matplotlib/ticker.py", line 2297, in tick_values
    "Data has no positive values, and therefore can not be "
ValueError: Data has no positive values, and therefore can not be log-scaled.
```

**Comment by mseitzer** [comment]
~This seems to be related to the plotting backend. When I do the following (where `'Qt5Agg' == plt.get_backend()`)~

```
self._cockpit_plotter._mpl_no_show_backend = 'Qt5Agg'
self._cockpit_plotter.plot(self._cockpit,
                                        savedir=cockpit_log_dir,
                                        show_plot=False,
                                        save_plot=True)
```

~everything works as expected.~

Seems like I was too quick, the crash still appears even when setting the backend.

**Comment by mseitzer** [comment]
I resolved the bug by updating matplotlib from `3.1.1` to the newest version. It appears that the way you use `matplotlib.use` to set the plotting backend was not supported in earlier versions, and thus matplotlib assumed an interactive mode by default. The interactive mode led to the `draw` function being called when it should not have been, which crashed (likely a bug in matplotlib itself). 

It might be sensible to require a minimum version of matplotlib when installing cockpit.

**Comment by fsschneider** [comment]
I see. Thanks for raising this issue.

I am currently using version `3.3.4` and I would be happy to declare a dependency of `matplotlib>=3.3.4` for `cockpit`.
What do you think @f-dangel ?

**Comment by f-dangel** [comment]
I'm locally on `3.3.4` as well. A quick search did not exactly reveal in which version that behavior was fixed; it could be [`3.3.3`](https://matplotlib.org/stable/users/prev_whats_new/whats_new_3.3.0.html?highlight=use#backend-specific-improvements), but I don't mind requiring `>=3.3.4`.

**Comment by fsschneider** [comment]
I decided to require `matplotlib>=3.4.0` since this version also fixes a different visual glitch for Apple Preview that occurred during one of the experiments we did for the paper.

Fixed with https://github.com/f-dangel/cockpit/commit/bb685e54fdfd54b02efe0aa8a01f3fe25258936d and https://github.com/f-dangel/cockpit/commit/db8ddc90590773dbf2b53daa8b19909ed74c5fb5
